Lib/test/test_call.py

```diff
@@ -457,6 +457,19 @@ def test_fastcall_keywords(self):
                 result = _testcapi.pyobject_fastcallkeywords(func, args, kwnames)
                 self.check_result(result, expected)
 
+    def test_fastcall_clearing_dict(self):
+        # Test bpo-36907: the point of the test is just checking that this
+        # does not crash.
+        class IntWithDict:
+            __slots__ = ["kwargs"]
+            def __init__(self, **kwargs):
+                self.kwargs = kwargs
+            def __index__(self):
+                self.kwargs.clear()
+                L = [2**i for i in range(10000)]
+                return 0
+        x = IntWithDict(dont_inherit=IntWithDict())
+        compile("pass", "", "exec", x, **x.kwargs)
```

Comment:
compile is quite a special function, so I recommend to add a note:

```suggestion
        # Note: we test compile's argument handling, not actual compilation
        compile("pass", "", "exec", x, **x.kwargs)
```